File: src/summary.py
"""Rule-based, strictly descriptive Czech summary of what changed.

Deterministic templates only — no LLM, so no interpretive or diagnostic
language can slip in. Describes direction, magnitude and any in/out-of-range
transition between an analyte's two most recent numeric results. Never says
what a change *means*.
"""
from __future__ import annotations

from typing import Optional

from .trends import Trend, TrendPoint, latest_two


def cz_num(x: Optional[float]) -> str:
    """Format a number Czech-style (decimal comma), trimming noise."""
    if x is None:
        return "—"
    if abs(x) >= 100:
        s = f"{x:.1f}"
    elif abs(x) >= 1:
        s = f"{x:.2f}"
    else:
        s = f"{x:.3f}"
    s = s.rstrip("0").rstrip(".") if "." in s else s
    return s.replace(".", ",")


def _range_str(low: Optional[float], high: Optional[float]) -> str:
    if low is not None and high is not None:
        return f"{cz_num(low)}–{cz_num(high)}"
    if high is not None:
        return f"< {cz_num(high)}"
    if low is not None:
        return f"> {cz_num(low)}"
    return "—"


def _direction(delta: float, eps: float) -> str:
    if delta > eps:
        return "vzrostlo"
    if delta < -eps:
        return "kleslo"
    return "beze změny"


def _range_transition(old_flag: str, new_flag: str) -> str:
    """Descriptive clause about crossing the reference range (or not)."""
    if new_flag == "high":
        return "a je nad referenčním rozmezím" if old_flag != "high" \
            else "a zůstává nad referenčním rozmezím"
    if new_flag == "low":
        return "a je pod referenčním rozmezím" if old_flag != "low" \
            else "a zůstává pod referenčním rozmezím"
    if new_flag == "normal":
        if old_flag in ("high", "low"):
            return "a je nyní v referenčním rozmezí"
        return "a je v referenčním rozmezí"
    return ""


def summarize_trend(trend: Trend) -> Optional[dict]:
    """One descriptive record for a trend's two most recent numeric results.

    Returns None when there aren't two numeric points to compare.
    """
    older, newer = latest_two(trend)
    if older is None or newer is None:
        return None

    unit = (trend.unit or "").strip()
    unit_sfx = f" {unit}" if unit else ""
    delta = newer.value - older.value
    # magnitude threshold: 1% of the older value (or tiny absolute) counts as change
    eps = max(abs(older.value) * 0.01, 1e-9)
    direction = _direction(delta, eps)

    if direction == "beze změny":
        text = (f"{trend.display_name} beze změny "
                f"({cz_num(older.value)} → {cz_num(newer.value)}{unit_sfx})")
    else:
        pct = (delta / older.value * 100) if older.value else None
        pct_part = f", {'+' if delta > 0 else '−'}{cz_num(abs(pct))} %" if pct is not None else ""
        text = (f"{trend.display_name} {direction} z {cz_num(older.value)} "
                f"na {cz_num(newer.value)}{unit_sfx} "
                f"({'+' if delta > 0 else '−'}{cz_num(abs(delta))}{pct_part})")

    transition = _range_transition(older.flag, newer.flag)
    if transition:
        rng = _range_str(newer.ref_low, newer.ref_high)
        rng_part = f" ({rng})" if rng != "—" else ""
        text += f" {transition}{rng_part}"
    text += "."

    # rank: out-of-range first, then bigger relative moves
    out_of_range = newer.flag in ("high", "low")
    rel = abs(delta) / abs(older.value) if older.value else 0.0
    return {
        "canonical_id": trend.canonical_id,
        "display_name": trend.display_name,
        "text": text,
        "out_of_range": out_of_range,
        "changed": direction != "beze změny",
        "new_flag": newer.flag,
        "older": older,
        "newer": newer,
        "rank": (0 if out_of_range else 1, -rel),
    }


def summarize_changes(trends: dict[str, Trend]) -> list[dict]:
    """Descriptive records for all comparable trends, most notable first."""
    records = [r for t in trends.values() if (r := summarize_trend(t)) is not None]
    records.sort(key=lambda r: r["rank"])
    return records

File: packages/lab-core/src/summary.ts
/**
 * Rule-based, strictly descriptive Czech summary of what changed.
 *
 * Ported from src/summary.py. Deterministic templates only — no LLM, so no
 * interpretive or diagnostic language can slip in. Describes direction,
 * magnitude and any reference-range transition between an analyte's two most
 * recent numeric results. Never says what a change *means*.
 */
import type { Flag } from "./models";
import { czDate, prettyUnit } from "./czech";
import { latestTwo, type Trend, type TrendPoint } from "./trends";

/**
 * Format a number Czech-style (decimal comma), trimming trailing noise.
 *
 * Rounds for display only. Where the printed value matters — a table beside
 * the source page, a trend row a reader may check against the document — use
 * `czExact`, because showing 3,8 for a printed 3,802 in a tool whose pitch is
 * verification against the source is exactly what makes a reader stop
 * trusting it.
 */
export function czNum(x: number | null | undefined): string {
  if (x === null || x === undefined || !Number.isFinite(x)) return "—";
  const a = Math.abs(x);
  let s = a >= 100 ? x.toFixed(1) : a >= 1 ? x.toFixed(2) : x.toFixed(3);
  if (s.includes(".")) s = s.replace(/0+$/, "").replace(/\.$/, "");
  return s.replace(".", ",");
}

/**
 * The value as printed, without rounding. Falls back to a formatted number
 * when there is no raw string to show (a converted unit, say).
 *
 * "As printed" minus the lab's own out-of-range marker: a "!" or "*" beside
 * the number is the lab's flag, which every screen here says in its own
 * words already. The stored valueRaw keeps it, so provenance is untouched.
 */
export function czExact(value: number | null | undefined, raw?: string | null): string {
  const printed = (raw ?? "").replace(/[!*]/g, "").trim();
  if (printed) return printed;
  if (value === null || value === undefined || !Number.isFinite(value)) return "—";
  // No rounding: render the number as-is with a Czech decimal comma.
  return String(value).replace(".", ",");
}

/**
 * A reference range as one string, the same on every screen.
 *
 * A lab that prints only an upper limit prints "< 4,14"; rendered as "–4,14"
 * that reads as a negative number, and a reader asked which side of it they
 * are on has to guess. Ondrej's call (2026-09-12): say "0–4,14" — the lower
 * limit that was implied all along, written down. An upper limit missing is
 * rarer and has no such implied number, so it keeps the lab's own "> 3,2".
 * Every screen formats a range through here; the sites used to each carry
 * their own template, and disagreed.
 */
export function czRange(low: number | null | undefined, high: number | null | undefined): string {
  const l = low ?? null;
  const h = high ?? null;
  if (l !== null && h !== null) return `${czNum(l)}–${czNum(h)}`;
  if (h !== null) return `0–${czNum(h)}`;
  if (l !== null) return `> ${czNum(l)}`;
  return "—";
}

function direction(delta: number, eps: number): "up" | "down" | "flat" {
  if (delta > eps) return "up";
  if (delta < -eps) return "down";
  return "flat";
}

/**
 * Clause about the reference range.
 *
 * Written without a verb on purpose. Czech verbs agree with the subject's
 * gender and number, and the subject here is an analyte name pulled from a
 * registry of 109 entries — "Triacylglyceroly vzrostlo" instead of "vzrostly"
 * is instantly visible to a native reader. Gendering the whole registry would
 * be a large, fragile change to fix phrasing; dropping the verb removes the
 * agreement problem entirely and reads like a lab report rather than prose.
 */
function rangeTransition(oldFlag: Flag, newFlag: Flag): string {
  if (newFlag === "high") return oldFlag !== "high" ? "nově nad rozmezím" : "stále nad rozmezím";
  if (newFlag === "low") return oldFlag !== "low" ? "nově pod rozmezím" : "stále pod rozmezím";
  if (newFlag === "normal")
    return oldFlag === "high" || oldFlag === "low" ? "nově v rozmezí" : "v rozmezí";
  return "";
}

export interface SummaryRecord {
  canonicalId: string;
  displayName: string;
  text: string;
  outOfRange: boolean;
  changed: boolean;
  newFlag: Flag;
  older: TrendPoint;
  newer: TrendPoint;
  rank: [number, number];
}

/** One descriptive record for a trend's two most recent numeric results. */
export function summarizeTrend(trend: Trend): SummaryRecord | null {
  const [older, newer] = latestTwo(trend);
  if (older === null || newer === null) return null;
  const ov = older.value as number;
  const nv = newer.value as number;

  const unit = prettyUnit(trend.unit);
  const unitSfx = unit ? ` ${unit}` : "";
  const delta = nv - ov;
  // magnitude threshold: 1% of the older value (or tiny absolute) counts
  const eps = Math.max(Math.abs(ov) * 0.01, 1e-9);
  const dir = direction(delta, eps);

  const sign = delta > 0 ? "+" : "−";
  const pct = ov ? (delta / ov) * 100 : null;
  // Whole percent: a doctor reads this as a magnitude, and "+16,67 %" implies
  // a precision that two measurements do not carry.
  const pctPart = pct !== null ? ` / ${sign}${Math.round(Math.abs(pct))} %` : "";

  // "beze změny" printed beside two visibly different numbers reads as a
  // contradiction and undermines the rest of the list, so a sub-threshold
  // move says so explicitly and still shows its size.
  const change =
    dir === "flat"
      ? `prakticky beze změny, ${sign}${czNum(Math.abs(delta))}${unitSfx}`
      : `${sign}${czNum(Math.abs(delta))}${unitSfx}${pctPart}`;

  // Dates included: a +17% move over six days and over six months are
  // different clinical facts, and the reader cannot tell them apart otherwise.
  let text =
    `${trend.displayName}: ${czNum(ov)} → ${czNum(nv)}${unitSfx} (${change})` +
    ` · ${czDate(older.date)} → ${czDate(newer.date)}`;

  const transition = rangeTransition(older.flag, newer.flag);
  if (transition) {
    const rng = czRange(newer.refLow, newer.refHigh);
    text += ` — ${transition}${rng !== "—" ? ` ${rng}` : ""}`;
  }

  // rank: out-of-range first, then bigger relative moves
  const outOfRange = newer.flag === "high" || newer.flag === "low";
  const rel = ov ? Math.abs(delta) / Math.abs(ov) : 0;
  return {
    canonicalId: trend.canonicalId,
    displayName: trend.displayName,
    text,
    outOfRange,
    changed: dir !== "flat",
    newFlag: newer.flag,
    older,
    newer,
    rank: [outOfRange ? 0 : 1, -rel],
  };
}

/** Descriptive records for all comparable trends, most notable first. */
export function summarizeChanges(trends: Map<string, Trend>): SummaryRecord[] {
  const records: SummaryRecord[] = [];
  for (const t of trends.values()) {
    const r = summarizeTrend(t);
    if (r) records.push(r);
  }
  records.sort((a, b) => a.rank[0] - b.rank[0] || a.rank[1] - b.rank[1]);
  return records;
}
